docs/okeps/okep-5224-connecting-udns/okep-5224-connecting-udns.md

```diff
@@ -0,0 +1,1675 @@
+    `TunnelKeysAllocator` will then allocate the new ones for the
+    connect router dynamically. This ensures we don't reuse
+    already allocated networkIDs. For each connect API that will
+    create a connect router, we will need 1 tunnel key per router.
+  * This value is annotated on the `ClusterNetworkConnect`
+    CRD for it to be then consumed on the `ovnkube-controller` side.
+    Sample:
+    ```yaml
+    k8s.ovn.org/connect-router-tunnel-key: '50000'
+    ```
+3) Status conditions will be updated based on feedback of setup
+ from the `ovnkube-controllers` in different zones.
+
+This controller also needs to watch for NADs to
+accommodate adds, updates and deletes for the network-subnet.
+
+**OVN-Kubernetes Controller**
+
+A new `network-connect-controller` will be created that watches the
+`ClusterNetworkConnect` objects and does the following:
+
+1) Based on the network selectors, creates the
+   necessary network topology changes mentioned
+   in the above section:
+   * Creates the `connect-router`
+   * Creates the ports connecting each Layer3 `ovn_cluster_router` and
+     each Layer2 `transit-router` to the `connect-router`
+   * Adds the necessary logical router policies on each network router
+   * Adds the necessary logical router static routes on the `connect-router`
+2) Reads the `k8s.ovn.org/network-subnets`
+   annotation set by the cluster manager on CNC and allocates
+   a `/31` point to point subnet for each layer3 node-network's router port link
+   and uses the IPs from the `/31` to create the ports. For layer2 we only
+   need one `/31` instead of per node /31's which is already allocated by
+   CM. This allocator will be a static allocator util that deterministically
+   always picks the same `/31` for every `node-id`. Example:
+   ```yaml
+    worker-1 (node-id=0) → 10.2.0.0/31 (10.2.0.0 + 10.2.0.1)
+    worker-2 (node-id=1) → 10.2.0.2/31 (10.2.0.2 + 10.2.0.3)
+    worker-3 (node-id=2) → 10.2.0.4/31 (10.2.0.4 + 10.2.0.5)
+    worker-4 (node-id=5) → 10.2.0.10/31 (10.2.0.10 + 10.2.0.11)
+   ```
+   We can see that if node-ids are not contiguous, then we will loose
+   those IPs in the middle. But we determined this is still better than
+   having annotation patch issues at scale. However this static allocator
+   totally relies on maxNodes value and node-id not exceeding the range
+   user has provided. Offset used will be `generator.GenerateP2PSubnet(nodeID * 2)`.
+   For layer2 we will just use the annotation value directly.
+3) Reads the `k8s.ovn.org/connect-router-tunnel-key` and uses
+   the tunnel key to create the `connect-router`
+4) Within each `connect-router`, we would need N_L3*M tunnel keys to
+   connect N_L3 Layer3 networks on M nodes and N_L2 tunnel keys to connect
+   N_L2 Layer2 networks on M nodes. So total of `N_L3*M+N_L2` keys.
+   * This will be done statically based on subnets allocated out for each
+     node-network pair for layer3 networks and for each layer2 network.
```

Comment:
I would like to see more detail here about how this will be calculated. I believe you guys figured out a plan right?

Reply:
OK Tim.

You make it sound like there are zero implementation details in the OKEP :) under implementation details.
and this specific point was caught on only last week when I raised it here: https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5246#discussion_r2431573110

Regardless, I don't find this a deal breaker. It's not much effort for me to include these details into the OKEP so let me do that so that you are happy. I'm not in any hurry.